asinfo.c

```diff
@@ -0,0 +1,255 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* asprintf() does not appear on linux without this */
+#define _GNU_SOURCE
+
+#include <arpa/inet.h>
+#include <arpa/nameser.h>
+#include <netinet/in.h>
+#include <errno.h>
+#include <netdb.h>
+#include <resolv.h>
+#include <stdlib.h>
+#include <string.h>
+#include "asinfo.h"
+#include "globals.h"
+
+static struct __res_state res;
+
+static const char *asinfo_from_ipv4(const char *, char **, char **);
+#ifdef asinfo_ipv6
+static const char *asinfo_from_ipv6(const char *, char **, char **);
+#endif
+static const char *asinfo_from_dns(const char *, char **, char **);
+static const char *keep_best(char **, char **, char *, char *);
+
+const char *
+asinfo_from_rr(const char *rrtype, const char *rdata,
+	       char **asnum, char **cidr)
+{
+	if (asinfo_lookup) {
+		if (strcmp(rrtype, "A") == 0)
+			return asinfo_from_ipv4(rdata, asnum, cidr);
+#ifdef asinfo_ipv6
+		if (strcmp(rrtype, "AAAA") == 0)
+			return asinfo_from_ipv6(rdata, asnum, cidr);
+#endif
+	}
+	return NULL;
+}
+
+bool
+asinfo_domain_exists(const char *domain) {
+	u_char buf[NS_PACKETSZ];
+
+	return res_query(domain, ns_c_in, ns_t_txt, buf, sizeof buf) > 0 ||
+		_res.res_h_errno != HOST_NOT_FOUND;
+}
+
+void
+asinfo_shutdown(void) {
+	if ((res.options & RES_INIT) != 0)
+		res_nclose(&res);
+}
+
+static const char *
+asinfo_from_ipv4(const char *addr, char **asnum, char **cidr) {
+	u_char a4[32/8];
+	char *dname;
+
+	if (inet_pton(AF_INET, addr, a4) < 0)
+		return strerror(errno);
+	int n = asprintf(&dname, "%d.%d.%d.%d.%s",
+			 a4[3], a4[2], a4[1], a4[0], asinfo_domain);
+	if (n < 0)
+		return strerror(errno);
+	const char *result = asinfo_from_dns(dname, asnum, cidr);
+	free(dname);
+	return result;
+}
+
+#ifdef asinfo_ipv6
+static const char *
+asinfo_from_ipv6(const char *addr, char **asnum, char **cidr) {
+	u_char a6[128/8];
+	const char *result;
+	char *dname, *p;
+	int i;
+
+	if (inet_pton(AF_INET6, addr, &a6) < 0)
+		return strerror(errno);
+	dname = malloc(strlen(asinfo_domain) + (128/4)*2);
+	if (dname == NULL)
+		return strerror(errno);
+	result = NULL;
+	p = dname;
+	for (i = (128/8) - 1; i >= 0; i--) {
+		int n = sprintf(p, "%x.%x.", a6[i] & 0xf, a6[i] >> 4);
+		if (n < 0) {
+			result = strerror(errno);
+			break;
+		}
+		p += n;
+	}
+	if (result == NULL) {
+		strcpy(p, asinfo_domain);
+		result = asinfo_from_dns(dname, asnum, cidr);
+	}
+	p = NULL;
+	free(dname);
+	return result;
+}
+#endif
+
+static const char *
```

Comment:
explain what we expect to be parsing.
why TXT>3 is an error because it doesn't match that expectation.

Reply:
To github.com:dnsdb/dnsdbq.git
   15c743d..36e9a8f  asnlookup -> asnlookup